plugin/health_check.go

```diff
@@ -12,105 +12,92 @@ import (
+	HEALTH_CHECK_INTERVAL            = 30 * time.Second // How often the health check should run
+	HEALTH_CHECK_DEACTIVATION_WINDOW = 60 * time.Minute // How long we wait for num fails to incur before deactivating the plugin
+	HEALTH_CHECK_PING_FAIL_LIMIT     = 3                // How many times we call RPC ping in a row before it is considered a failure
+	HEALTH_CHECK_RESTART_LIMIT       = 3                // How many times we restart a plugin before we deactivate it
 )
 
-type PluginHealthCheckJob struct {
-	cancel     chan struct{}
-	cancelled  chan struct{}
-	cancelOnce sync.Once
-	env        *Environment
+type PluginHealthCheckStatus struct {
+	failTimestamps []time.Time
+	Crashed        bool
 }
 
-// InitPluginHealthCheckJob starts a new job if one is not running and is set to enabled, or kills an existing one if set to disabled.
-func (env *Environment) InitPluginHealthCheckJob(enable bool) {
-	// Config is set to enable. No job exists, start a new job.
-	if enable && env.pluginHealthCheckJob == nil {
-		mlog.Debug("Enabling plugin health check job", mlog.Duration("interval_s", HEALTH_CHECK_INTERVAL))
-
-		job := newPluginHealthCheckJob(env)
-		env.pluginHealthCheckJob = job
-		job.Start()
-	}
-
-	// Config is set to disable. Job exists, kill existing job.
-	if !enable && env.pluginHealthCheckJob != nil {
-		mlog.Debug("Disabling plugin health check job")
-
-		env.pluginHealthCheckJob.Cancel()
-		env.pluginHealthCheckJob = nil
-	}
+type PluginHealthCheckJob struct {
+	cancel              chan struct{}
+	cancelled           chan struct{}
+	cancelOnce          sync.Once
+	env                 *Environment
+	healthCheckStatuses sync.Map
 }
 
 // Start continuously runs health checks on all active plugins, on a timer.
 func (job *PluginHealthCheckJob) Start() {
 	mlog.Debug("Plugin health check job starting.")
+	defer close(job.cancelled)
 
-	go func() {
-		defer close(job.cancelled)
-
-		ticker := time.NewTicker(HEALTH_CHECK_INTERVAL)
-		defer func() {
-			ticker.Stop()
-		}()
-
-		for {
-			select {
-			case <-ticker.C:
-				activePlugins := job.env.Active()
-				for _, plugin := range activePlugins {
-					job.checkPlugin(plugin.Manifest.Id)
-				}
-			case <-job.cancel:
-				return
+	ticker := time.NewTicker(HEALTH_CHECK_INTERVAL)
+	defer func() {
+		ticker.Stop()
+	}()
+
+	for {
+		select {
+		case <-ticker.C:
+			activePlugins := job.env.Active()
+			for _, plugin := range activePlugins {
+				job.CheckPlugin(plugin.Manifest.Id)
 			}
+		case <-job.cancel:
+			return
 		}
-	}()
+	}
 }
 
-// checkPlugin determines the plugin's health status, then handles the error or success case.
-func (job *PluginHealthCheckJob) checkPlugin(id string) {
-	p, ok := job.env.registeredPlugins.Load(id)
+// CheckPlugin determines the plugin's health status, then handles the error or success case.
+func (job *PluginHealthCheckJob) CheckPlugin(id string) {
+	statusInterface, ok := job.healthCheckStatuses.Load(id)
```

Comment:
if the load check is moved from here to handleHealthCheckFail function then this whole function goes away as it'll only be:
```
	err := job.env.performHealthCheck(id)
	if err != nil {
		mlog.Error("Health check failed for plugin", mlog.String("id", id), mlog.Err(err))
		job.handleHealthCheckFail(id, status)
	}
```
which can be moved inside `Start()` Thoughts?

Reply:
This currently won't work with the current state of using the `Crashed` boolean, because the `Crashed` value will not be updated to `false` upon externally re-activating the plugin. The `HealthCheckStatus`'s `Crashed` value must be updated when we know that the plugin is currently in a state where we can check its health. i.e. it has not failed 3 times.

Reply:
@ali-farooq0 ^ I think this is the last main piece of feedback that needs to be resolved. The current state of the code that you point out is not too elegant, but we need to make sure that the healthiness (`Crashed = false`) is reset when the plugin is back online.